49-hvac-current-monitor/firmware/src/main.cpp

```diff
@@ -0,0 +1,330 @@
+            // variables.
+            sendMeasurementsNote(measurements, measurementsIdx,
+                msBetweenMeasurements, pctChangeThreshold,
+                &lastAvgCurrent);
+
+            msBetweenMeasurements = seconds * 1000; // Convert seconds to ms.
+            Serial.print("Seconds between measurements set to ");
+            Serial.print(seconds);
+            Serial.println(".");
+        }
+    }
+    else if (strcmp(var, "percentage_change_threshold") == 0) {
+        float threshold = atof(val);
+
+        if (threshold <= 0) {
+            Serial.println("percentage_change_threshold must be a positive "
+                "number.");
+        }
+        else {
+            // Flush out the measurements before changing any environment
+            // variables.
+            sendMeasurementsNote(measurements, measurementsIdx,
+                msBetweenMeasurements, pctChangeThreshold,
+                &lastAvgCurrent);
+
+            pctChangeThreshold = threshold;
+            Serial.print("Percentage change threshold set to ");
+            Serial.print(pctChangeThreshold);
+            Serial.println(".");
+        }
+    }
+}
+
+float calcVrms()
+{
+    static float samples[SAMPLES_PER_CALC];
+
+    int sample = 0;
+    float sum = 0;
+
+    // Read the samples into an array and sum them up so we can compute the DC
+    // offset via an average.
+    for (size_t i = 0; i < SAMPLES_PER_CALC; ++i) {
+        samples[i] = analogRead(ANALOG_PIN);
+        sum += samples[i];
+    }
+
+    // Average the samples to get the DC offset.
+    float offset = sum / SAMPLES_PER_CALC;
+    float sampleWithOffsetRemoved = 0;
+    sum = 0;
+
+    for (size_t i = 0; i < SAMPLES_PER_CALC; ++i) {
+        // Subtract the offset from each sample.
+        sampleWithOffsetRemoved = samples[i] - offset;
+        // Sum the squares of the samples with offset removed.
+        sum += sampleWithOffsetRemoved * sampleWithOffsetRemoved;
+    }
+
+    // Use the sum of squares to compute the RMS voltage.
+    float rms = (sqrt(sum / SAMPLES_PER_CALC) / ADC_RANGE) * ADC_REF_VOLTAGE;
+
+    return rms;
+}
+
+// Register the template for current consumption data.
+void registerTemplate()
+{
+    J *req = notecard.newRequest("note.template");
+    if (req == NULL) {
+        Serial.println("Failed to create note.template request.");
+        return;
+    }
+
+    J *body = JCreateObject();
+    if (body == NULL) {
+        JDelete(req);
+        Serial.println("Failed to create note.template request body.");
+        return;
+    }
+
+    JAddStringToObject(req, "file", MEASUREMENTS_FILE);
+    JAddNumberToObject(body, "seconds_between_measurements", TUINT32);
+    JNUMBER measurementsTemplate[MEASUREMENTS_PER_NOTE];
+    for (size_t i = 0; i < MEASUREMENTS_PER_NOTE; ++i) {
+        measurementsTemplate[i] = TFLOAT32;
+    }
+    J *data = JCreateNumberArray(measurementsTemplate, MEASUREMENTS_PER_NOTE);
```

Comment:
iirc array templates are populated with 0's when the template isn't full? Not a problem, with the `length` field.

Reply:
I believe that's correct. Because we can get a "premature" flush from environment variable changes, I had to include the length field to indicate how much of the data is valid.